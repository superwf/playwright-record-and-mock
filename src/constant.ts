import { Config } from './type'

export const PKG_NAME = 'playwright-record-and-mock'
export const PLAYWRIGHT_CONFIG_FILE = 'playwright.config.ts'

export const TEST_CASE_FILE_NAME = 'run.spec.ts'
export const MAIN_FIXTURE_FILE = 'fixture.json'
export const FIXTURES_DIR = 'fixtures'

export const DEFAULT_VIEWPORT = { width: 1920, height: 974 }

export const DEFAULT_CONFIG: Config = {
  outDir: 'e2e',
  site: 'https://cn.bing.com/',
  urlFilter: /\/AS\/Suggestions/,
}
