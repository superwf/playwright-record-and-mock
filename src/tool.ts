import type { Request, BrowserContextOptions } from '@playwright/test'
import { emptydirSync } from 'fs-extra'
import minimatch from 'minimatch'
import path from 'path'
import { UrlFilter } from './type'
import { MAIN_FIXTURE_FILE, TEST_CASE_FILE_NAME, FIXTURES_DIR } from './constant'

export const resolveRoot = (...paths: string[]) => path.resolve(process.cwd(), path.join(...paths))

export const encodeToBase64 = (str: string | Buffer): string => Buffer.from(str).toString('base64')

export const decodeFromBase64 = (base64str: string): Buffer => Buffer.from(base64str, 'base64')

export const isContentTypeText = (contentType: string) => /text|script|xml|xhtml/.test(contentType)
export const isContentTypeJson = (contentType: string) => /application\/json/.test(contentType)

export const viewportSizeToViewportDimension = (
  viewport?: string | BrowserContextOptions['viewport'],
): BrowserContextOptions['viewport'] => {
  if (typeof viewport === 'string') {
    if (!/^\d+,\d+$/.test(viewport)) {
      throw new Error('viewportSize must be dimension like "1920,1080"')
    }
    const [width, height] = viewport.split(',').map<number>(n => parseInt(n, 10))
    return {
      width,
      height,
    }
  }
  return viewport
}

export const getTestCaseFilePath = (outDir: string, caseName: string) =>
  resolveRoot(path.join(outDir, caseName, TEST_CASE_FILE_NAME))

export const getTestCaseMainFixtureFilePath = (outDir: string, caseName: string) =>
  resolveRoot(path.join(outDir, caseName, MAIN_FIXTURE_FILE))

export const getTestCaseFixtureFilePath = (outDir: string, caseName: string, dataFile: string) =>
  resolveRoot(path.join(outDir, caseName, FIXTURES_DIR, dataFile))

export const cleanTestCaseFixtureFilePath = (outDir: string, caseName: string) => {
  const dir = resolveRoot(path.join(outDir, caseName, FIXTURES_DIR))
  emptydirSync(dir)
}

export const isUrlMatched = (url: URL, urlFilter?: UrlFilter): boolean => {
  if (!urlFilter) {
    return true
  }
  if (typeof urlFilter === 'string') {
    return minimatch(url.href, urlFilter)
  }
  if (urlFilter instanceof RegExp) {
    return urlFilter.test(url.href)
  }
  if (typeof urlFilter === 'function') {
    return urlFilter(url)
  }
  throw new Error('urlFilter must be string, regexp or function')
}

export const generateResponseMapKey = async (req: Request) => {
  const url = new URL(req.url())
  return `${req.method()}+${url.protocol}//${url.host}${url.pathname}`
}

export const sleep = (n: number) =>
  new Promise(resolve => {
    setTimeout(() => resolve(null), n)
  })
