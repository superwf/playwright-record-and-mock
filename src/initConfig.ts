import fs from 'fs'
import { parse } from '@babel/parser'
import traverse from '@babel/traverse'
import generate from '@babel/generator'
import template from '@babel/template'
import * as t from '@babel/types'
import { resolveRoot } from './tool'
import { PLAYWRIGHT_CONFIG_FILE, DEFAULT_CONFIG, PKG_NAME } from './constant'
import { Config } from './type'
import { ok, log, err } from './logger'

export const initConfig = (config?: Config) => {
  config = config || DEFAULT_CONFIG
  const playwrightConfigFile = resolveRoot(PLAYWRIGHT_CONFIG_FILE)
  if (!fs.existsSync(playwrightConfigFile)) {
    throw new Error(
      `you shoul prepare "${PLAYWRIGHT_CONFIG_FILE}" first! Go to read https://playwright.dev/docs/test-configuration#global-configuration`,
    )
  }
  const source = fs.readFileSync(playwrightConfigFile, 'utf8')
  const ast = parse(source, {
    sourceType: 'module',
    plugins: ['typescript'],
  })
  let injected = false
  traverse(ast, {
    ImportDeclaration(path) {
      if (path.node.source.value === PKG_NAME) {
        injected = true
      }
    },
  })
  if (injected) {
    log(err(`your ${PLAYWRIGHT_CONFIG_FILE} already injected ${PKG_NAME} config`))
    throw new Error(`can not reinject ${PKG_NAME} config`)
  }

  traverse(ast, {
    ImportDeclaration(path) {
      if (path.node.source.value === '@playwright/test') {
        const templ = template.ast(`import { PramConfig } from 'playwright-record-and-mock'`)
        path.insertAfter(templ)
        path.skip()
      }
    },
    TSTypeAnnotation(path) {
      const { typeAnnotation } = path.node
      if (t.isTSTypeReference(typeAnnotation)) {
        const { typeName } = typeAnnotation
        if (t.isIdentifier(typeName)) {
          if (typeName.name === 'PlaywrightTestConfig') {
            // path.insertAfter(t)
            const n = t.tsIntersectionType([
              t.tsTypeReference(t.identifier('PlaywrightTestConfig')),
              t.tsTypeReference(t.identifier('PramConfig')),
            ])
            const n1 = t.tsTypeAnnotation(n)
            path.replaceWith(n1)
            const idenPath = path.parentPath?.parentPath
            const identityNode = idenPath?.node
            if (idenPath && t.isVariableDeclarator(identityNode)) {
              const { init } = identityNode
              if (t.isObjectExpression(init)) {
                if (
                  init.properties.some(p => {
                    if (t.isObjectProperty(p) && t.isIdentifier(p.key) && p.key.name === 'pram') {
                      return true
                    }
                    return false
                  })
                ) {
                  return
                }
                const configNode = t.objectExpression(
                  Object.getOwnPropertyNames(config).map<t.ObjectProperty>(key => {
                    const value = config![key as keyof Config]
                    if (typeof value === 'string') {
                      return t.objectProperty(t.identifier(key), t.stringLiteral(value))
                    }
                    if (value instanceof RegExp) {
                      return t.objectProperty(t.identifier(key), t.regExpLiteral(String(value).slice(1, -1)))
                    }
                    // istanbul ignore next
                    throw new Error(`unknown config property type, key: key, value: ${value}`)
                  }),
                )
                // [
                // t.objectProperty(t.identifier('outDir'), t.stringLiteral('e2e')),
                // t.objectProperty(t.identifier('site'), t.stringLiteral('https://cn.bing.com/')),
                // t.objectProperty(t.identifier('urlFilter'), t.regExpLiteral('\\/AS\\/Suggestions')),
                // ])
                init.properties.push(t.objectProperty(t.identifier('pram'), configNode))
              }
              idenPath.replaceWith(idenPath.node)
            }
            path.skip()
          }
        }
      }
    },
  })
  const newCode = generate(ast).code
  fs.writeFileSync(playwrightConfigFile, newCode)
  log('write to', ok(PLAYWRIGHT_CONFIG_FILE), 'success.')
}
