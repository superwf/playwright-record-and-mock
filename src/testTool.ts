import temp from 'temp'
import { resolveRoot } from './tool'

/**
 * 创建一个临时文件夹作为当前目录
 * 完成后还原目录并清理临时文件
 * 仅测试用
 * */
export const testInTempPath = () => {
  const testPath = temp.mkdirSync(resolveRoot('temp-path-for-test'))
  const cwd = process.cwd()
  process.chdir(testPath)
  return {
    testPath,
    restore() {
      temp.cleanupSync()
      process.chdir(cwd)
    },
  }
}
