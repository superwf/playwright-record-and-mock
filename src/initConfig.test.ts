import fs from 'fs'
import path from 'path'
import { initConfig } from './initConfig'
import { PLAYWRIGHT_CONFIG_FILE } from './constant'
import { resolveRoot } from './tool'
import { testInTempPath } from './testTool'

test('create initConfig file', () => {
  const originConfigFile = resolveRoot(PLAYWRIGHT_CONFIG_FILE)
  const { testPath, restore } = testInTempPath()
  const configTargetFile = path.join(testPath, PLAYWRIGHT_CONFIG_FILE)
  fs.cpSync(originConfigFile, configTargetFile)
  expect(fs.existsSync(configTargetFile)).toBe(true)
  initConfig()
  const code = fs.readFileSync(resolveRoot(PLAYWRIGHT_CONFIG_FILE), {
    encoding: 'utf8',
  })
  expect(code).toMatchSnapshot()

  // throw if rerun init
  expect(initConfig).toThrow()
  restore()
})

test('if not exist playwrightConfigFile, throw', () => {
  const { restore } = testInTempPath()
  expect(initConfig).toThrow()
  restore()
})
