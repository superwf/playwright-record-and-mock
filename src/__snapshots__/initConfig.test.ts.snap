// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create initConfig file 1`] = `
"import type { PlaywrightTestConfig } from '@playwright/test';
import { PramConfig } from 'playwright-record-and-mock';
const config: PlaywrightTestConfig & PramConfig = {
  projects: [{
    name: 'Chrome Stable',
    testDir: 'e2e'
  }],
  use: {
    headless: false,
    browserName: 'chromium',
    channel: 'chrome',
    viewport: {
      width: 1920,
      height: 1080
    }
  },
  pram: {
    outDir: \\"e2e\\",
    site: \\"https://cn.bing.com/\\",
    urlFilter: /\\\\/AS\\\\/Suggestions/
  }
};
export default config;"
`;
